### ISSUE TITLE
Truncates timestamps when timezone utc offset has seconds

### ISSUE THREAD

**Issue by olasd** [psycopg/psycopg2#1272]
Hi,

Thank you for maintaining psycopg2!

While working on historical timestamps stored in PostgreSQL, we've tracked down a loss of precision, when using a timezone with a UTC offset containing a non-zero number of seconds (for instance, Europe/Paris around 1777).

Here's a script showing the issue (I've run it on PostgreSQL 13.2 running on Debian sid - with tzdata 2021a) :

```python
#!/usr/bin/env python3

import psycopg2

db = psycopg2.connect("")
with db.cursor() as cur:
    cur.execute("create temp table test (ts timestamptz) on commit drop;")
    cur.execute("insert into test values (%s)", ("1777-11-13 18:30:00+00", ))

    results = {}
    text = {}
    for timezone in ("UTC", "Europe/Paris"):
        cur.execute("set timezone to %s", (timezone, ))
        cur.execute("select ts from test")
        results[timezone] = cur.fetchone()[0]
        cur.execute("select ts::text from test")
        text[timezone] = cur.fetchone()[0]

for timezone in results:
    ts = results[timezone]
    print(timezone, ts, ts.timestamp(), text[timezone])
```

The output is the following:
```
UTC 1777-11-13 18:30:00+00:00 -6063053400.0 1777-11-13 18:30:00+00
Europe/Paris 1777-11-13 18:39:21+00:09 -6063053379.0 1777-11-13 18:39:21+00:09:21
```

In the second line, you can see the fractional UTC offset for Europe/Paris (0 hours, 9 minutes and 21 seconds).

You can clearly see that the timestamp values retrieved by psycopg2 are different, even though they should be the same. I believe psycopg2 parses the timezone specification returned by PostgreSQL, and truncates the seconds.

I'm happy to track this down and propose a fix if this is to be considered a bug.

**Comment by dvarrazzo** [comment]
Yes, this is documented: https://www.psycopg.org/docs/usage.html?highlight=timezone#time-zones-handling

I am of the understanding that Python doesn't handle seconds in the timezones. This was definitely the case in the past. Has this improved?

**Comment by dvarrazzo** [comment]
I am happening to be working on psycopg3 datetime handling and stumbled again in this issue. An observation: returning datetimetz always in utc would make the problem of the seconds in the timezone offset disappear. Considering to do that in psycopg3 (not because of the Python 3.6 problem, rather because in binary format the tzinfo is not available anyway: it is a connection property, not a data property).

**Comment by asqui** [comment]
Hmmm, are you saying that `timestamptz` values from the database would always be returned in Python as a `datetime` instance with `tzinfo=UTC`?

If I have understood that correctly, that seems like a pretty big breaking change, at least for our codebase.

How about tacking on an `.astimezone(connection_tz)` to do that conversion on the client, or is that too naive?

**Comment by dvarrazzo** [comment]
> If I have understood that correctly, that seems like a pretty big breaking change, at least for our codebase.

Interesting. The two timestamps are equivalent (except when there is an error like here of course): why would it be a breaking change?

> How about tacking on an .astimezone(connection_tz) to do that conversion on the client, or is that too naive?

I can think about that. It sounds a lot of work (computationally speaking) to get an equivalent result so I wonder if it can be avoided.

**Comment by dvarrazzo** [comment]
In order to parse the connection tz info we would need to depend on `pytz`. Not terribly bad: it's a pretty standard library. However event that doesn't seem to manage all the timezones postgres does. For instance "+02:00" is a valid Postgres timezone but `pytz.timezone("+02:00")` raises `UnknownTimeZoneError`.

Returning all timestamptz in UTC is not a thing I would do in psycopg2 but I think it would be an ok change in psycopg3. Happy to know if you have use cases showing how terrible an idea it is, @asqui :)

**Comment by asqui** [comment]
Switching to return all timestamptz in UTC means that psycopg2 code which was getting results in the connection timezone is now going to get them in UTC after upgrading to psycopg3.

Why does this matter?

In psycopg2 you can rely on results being in the session timezone; with the proposed change you have to call `.astimezone()` on all results, because the session timezone is ignored. It's not the end of the world, but it's a change.

I understand the academic argument that the UTC datetime refers to the same _instant_ as what was being output before, but the practical argument is that the calling code might be doing any number of things that rely on the timezone.

Some examples:
1. Calling code outputs the result directly via `isoformat()` e.g. by formatting the timestamp into a string, a web page, a CSV file. Yes the timestamp refers to the same _instant_, but upgrading to psycopg3 has changed the output of my routine. Maybe I need to output in a particular timezone, and the way I was achieving that was by setting the session timezone (or relying on knowing what the server default is).
2. Worse, maybe the calling code formats the result into a particular format: `.strftime('%Y-%m-%d %H:%M:%S')` Whoops, we just lost the TZ information. But it's okay, because we _know_ that it's in the session timezone, right? Well, we knew that with psycopg2. With the proposed change, we now have a bug.

I think such a change would make it incredibly difficult to switch any moderately sized codebase that deals with databases with a known default timezone, or relies on setting the session timezone, from psycopg2 to 3 without the risk  of a long tail of subtle bugs around this.

It would also be a surprise, because the behaviour of results retrieved with psycopg3 will differ from what you see executing interactive queries with psql, which I think will be incredibly confusing to users trying to debug and investigate their python code and data.

Hope that makes sense?

**Comment by dvarrazzo** [comment]
It makes sense. I don't think it would be so much of a deal breaker for porting, but I can see how it can become a cause of frustration.

Moving this conversation to https://github.com/psycopg/psycopg3/discussions/56 - thank you @asqui for the conversation as usually constructive :)

As for the state of things in psycopg2, I think that the correct fix would be not only to use the seconds in the fixed offset timezone in Python 3.7, on but also to move the error from the timestamp to the timezone offset, as suggested in the above psycopg3 conversation (Not that Python 3.6 has still a long lifetime). Does it make sense for you, @olasd ?

**Comment by dvarrazzo** [comment]
The issue will be fixed in the upcoming psycopg 2.9 release. In such release the `datetime.timezone` object is used as default `cursor.tzinfo_factory` instead of `psycopg2.tz.FixedOffsetTimezone`. The object is still available in case someone has subclassed it, but will be dropped in the next major release.
